### PR TITLE
Add versioned git-hooks

### PR DESCRIPTION
**Pull request by midwire**
* Should remove git-hooks formula from head-only.

### CODE REVIEW

File: Library/Formula/git-hooks.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+require "formula"
+
+class GitHooks < Formula
+  homepage "https://github.com/icefox/git-hooks"
+  head "https://github.com/icefox/git-hooks.git"
```
Comment (mikemcquaid) [inline Library/Formula/git-hooks.rb]: Use `https://github.com/icefox/git-hooks/archive/1.00.0.tar.gz` as the `url` and add a `sha1`.

File: Library/Formula/git-hooks.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+  def install
+    bin.install "git-hooks"
+  end
+end
```
Comment (mikemcquaid) [inline Library/Formula/git-hooks.rb]: Can a test be added to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

File: Library/Formula/git-hooks.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class GitHooks < Formula
+  homepage "https://github.com/icefox/git-hooks"
+  url "https://github.com/icefox/git-hooks/archive/1.00.0.tar.gz"
+  version "1.00.0"
```
Comment (jawshooah) [inline Library/Formula/git-hooks.rb]: `brew audit --strict` doesn't like this. The version is already scanned from the `url`.

File: Library/Formula/git-hooks.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      Run 'git hooks --install' Inside a git project to change that projects
+      git hooks to use git-hooks hooks. 'git hooks --uninstall' can be run at
+      any time to go back to the old hooks that were installed before
+      (typically the default which does nothing).
+
+      Run 'git hooks --installglobal' to force any new git repository or any
+      git repository you clone to have a reminders to install git hooks. (It
+      can't be on by default for security reasons of course)
+    EOS
+  end
```
Comment (mikemcquaid) [inline Library/Formula/git-hooks.rb]: Let's kill these caveats. People will use the README instead.

File: Library/Formula/git-hooks.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+class GitHooks < Formula
+  homepage "https://github.com/icefox/git-hooks"
+  url "https://github.com/icefox/git-hooks/archive/1.00.0.tar.gz"
+  version "1.00.0"
+  sha1 "60b6048b8442237914f6174113facb922a9a3cdd"
+
+  def install
+    bin.install "git-hooks"
```
Comment (jawshooah) [inline Library/Formula/git-hooks.rb]: Maybe add `(etc/"git-hooks").install "contrib"` here as well?

**Comment by nathanhruby** [comment]
Are these review comments blockers?  If so, @midwire, would you accept a patch from me to get this moving?

**Comment by midwire** [comment]
Hey @nathanhruby, I just haven't had time to get to it (see: day job). Please feel free to proceed.

**Comment by nathanhruby** [comment]
added midwire/homebrew#1 since I have no idea how to update someone elses PR in Github

**Comment by nathanhruby** [comment]
Added midwire/homebrew#2 to address bot failures, also ran the formula though a more current "brew audit --strict" and it passes cleanly

**Comment by mistydemeo** [comment]
Can you squash this down to one commit, please?

File: Library/Formula/git-hooks.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  url "https://github.com/icefox/git-hooks/archive/1.00.0.tar.gz"
+  sha256 "8197ca1de975ff1f795a2b9cfcac1a6f7ee24276750c757eecf3bcb49b74808e"
+
+  def install
+    bin.install "git-hooks"
+    (etc/"git-hooks").install "contrib"
+  end
+
+  test do
+    output = `cd $(brew --prefix); git hooks`.strip
```
Comment (mistydemeo) [inline Library/Formula/git-hooks.rb]: Nitpick, but `brew --repository`, not `brew --prefix`

Comment (mistydemeo) [inline Library/Formula/git-hooks.rb]: In addition, you can get this value from the `HOMEBREW_REPOSITORY` variable:

```ruby
cd HOMEBREW_REPOSITORY do
  assert_match /Listing User, Project, and Global hooks/, shell_output("git hooks")
end
```

**Comment by nathanhruby** [comment]
Thanks @midwire!  

@mistydemeo I think this is good to go.  If the note about HOMEBREW_REPOSITORY is something that needs fixed, but not critical for merge, can we merge and then fix so I can stop bugging @midwire?

Thanks!

**Comment by mikemcquaid** [comment]
Thanks both of you for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by DomT4** [comment]
@mikemcquaid What's the mechanism for moving stuff from taps to core? We seem to handle moving stuff from core to taps okay, rarely an error there, but the recent ` x11 => core ` and ` devel-only => core ` stuff has raised something similar to this error:

```
Warning: Could not create link for homebrew/head-only/git-hooks, as it
conflicts with Homebrew/homebrew/git-hooks. You will need to use the
fully-qualified name when referring this formula, e.g.
  brew install homebrew/head-only/git-hooks
```

Even after those formulae are removed from the tap in question. Can we do anything on this front/Do we need to do anything on this front? It doesn't routinely reproduce across updates, just the initial one.

**Comment by mikemcquaid** [comment]
@DomT4 Hmm, I guess it should be handled but isn't currently? I thought when you `brew update` it untapped and retapped everything.

**Comment by DomT4** [comment]
Yeah, I presume [this block](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/update.rb#L19-L33) and particularly [this block](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/update.rb#L55-L57) is supposed to be untapping and retapping everything.

It looks like we're updating Homebrew/homebrew *before* updating and untapping and retapping ` tap `s. Maybe moving the tap-formula untapping up before the Homebrew/homebrew update would solve things? Unless I'm misreading that block.

**Comment by mikemcquaid** [comment]
@DomT4 Yeh, untapping first would make sense to me as long as we handle the error-case when the update fails and retap.

**Comment by DomT4** [comment]
I'm trying to work out the best way to test this, given we can't move things in and out of Homebrew/homebrew constantly just for the sake of getting reproducible tests :smile_cat:. Could always create a fake-fork and fake-tap and then wiggle them around locally until the errors stop appearing, I guess - Unless you're happy to bash out a solution here and then merge it without much direct testing?

How about this? Any good? Awful? Useless? :wink: 

```diff
diff --git a/Library/Homebrew/cmd/update.rb b/Library/Homebrew/cmd/update.rb
index 6e065b1..20f4d4d 100644
--- a/Library/Homebrew/cmd/update.rb
+++ b/Library/Homebrew/cmd/update.rb
@@ -16,12 +16,14 @@ module Homebrew
     cd HOMEBREW_REPOSITORY
     git_init_if_necessary
 
+    # we unlink first in case the formula has moved to another tap
     tapped_formulae = []
     HOMEBREW_LIBRARY.join("Formula").children.each do |path|
       next unless path.symlink?
       tapped_formulae << path.resolved_path
     end
     unlink_tap_formula(tapped_formulae)
+    Homebrew.unlink_tap_formula(report.removed_tapped_formula)
 
     report = Report.new
     master_updater = Updater.new(HOMEBREW_REPOSITORY)
@@ -43,19 +45,17 @@ module Homebrew
         begin
           updater.pull!
         rescue
+          Homebrew.link_tap_formula(report.new_tapped_formula)
           onoe "Failed to update tap: #{user.basename}/#{repo.basename.sub("homebrew-", "")}"
         else
           report.update(updater.report) do |key, oldval, newval|
             oldval.concat(newval)
           end
+        Homebrew.link_tap_formula(report.new_tapped_formula)
         end
       end
     end
 
-    # we unlink first in case the formula has moved to another tap
-    Homebrew.unlink_tap_formula(report.removed_tapped_formula)
-    Homebrew.link_tap_formula(report.new_tapped_formula)
-
     # automatically tap any migrated formulae's new tap
     report.select_formula(:D).each do |f|
       next unless (HOMEBREW_CELLAR/f).exist?
```

**Comment by mikemcquaid** [comment]
Something like that makes sense but maybe we just call out to `brew tap --repair`.

**Comment by DomT4** [comment]
Presumably between the Homebrew/homebrew pull and the taps pull?

**Pull request closed without merging**